org.openscenegraph.android/jni/JNIosg.cpp

```diff
@@ -139,6 +178,931 @@ JNIEXPORT jint JNICALL Java_org_openscenegraph_osg_core_Group_nativeGetNumChildr
+{
+	osg::Geometry *g = reinterpret_cast<osg::Geometry*>(cptr);
+	g->setColorBinding(osg::Geometry::AttributeBinding(binding));
+}
+
+JNIEXPORT jboolean JNICALL Java_org_openscenegraph_osg_core_Geometry_nativeAddPrimitiveSet(JNIEnv *, jclass, jlong cptr, jlong set_cptr)
+{
+	osg::Geometry *g = reinterpret_cast<osg::Geometry*>(cptr);
+	osg::PrimitiveSet *n = reinterpret_cast<osg::PrimitiveSet*>(set_cptr);
+	if(g!=NULL)
+	{
+        return g->addPrimitiveSet(n)?JNI_TRUE:JNI_FALSE;
+	}
+	return JNI_FALSE;
+}
+
+JNIEXPORT void JNICALL Java_org_openscenegraph_osg_core_Geometry_nativeSetPrimitiveSetList(JNIEnv* env, jclass, jlong cptr, jlong list_cptr)
+{
+	osg::Geometry* g = reinterpret_cast<osg::Geometry*>(cptr);
+	osg::Geometry::PrimitiveSetList* n = reinterpret_cast<osg::Geometry::PrimitiveSetList*>(list_cptr);
+	if(g!=NULL)
+	{
+		g->setPrimitiveSetList(*n);
+	}
+}
+
+/*
+ * osg::Drawable
+ */
+JNIEXPORT void JNICALL Java_org_openscenegraph_osg_core_Drawable_nativeDispose(JNIEnv* env, jclass, jlong cptr)
+{
+	osg::Drawable* g = reinterpret_cast<osg::Drawable*>(cptr);
+	if(g!=NULL)
+		g->unref();
+}
+
+JNIEXPORT jlong JNICALL Java_org_openscenegraph_osg_core_Drawable_nativeAsDrawable(JNIEnv *, jclass, jlong cptr)
+{
+	osg::Drawable *g = reinterpret_cast<osg::Drawable *>(cptr);
+	return reinterpret_cast<jlong>(g);
+}
+
+/*
+ * osg::PrimitiveSet
+ */
+JNIEXPORT void JNICALL Java_org_openscenegraph_osg_core_PrimitiveSet_nativeDispose(JNIEnv *, jclass, jlong cptr)
+{
+	osg::PrimitiveSet *g = reinterpret_cast<osg::PrimitiveSet *>(cptr);
+	if(g!=NULL)
+		g->unref();
+}
+
+/*
+JNIEXPORT jlong JNICALL Java_org_openscenegraph_osg_core_PrimitiveSet_nativeCreatePrimitiveSet(JNIEnv *, jclass, jint type, jint mode)
+{
+	osg::PrimitiveSet *g = NULL;
+	switch(type)
+	{
+	case 0:
+	{
+		g = new osg::PrimitiveSet(osg::PrimitiveSet::PrimitiveType, (int)mode);
+		break;
+	}
+	case 1:
+	{
+		g = new osg::PrimitiveSet(osg::PrimitiveSet::DrawArraysPrimitiveType, (int)mode);
+		break;
+	}
+	case 2:
+	{
+		g = new osg::PrimitiveSet(osg::PrimitiveSet::DrawArrayLengthsPrimitiveType, (int)mode);
+		break;
+	}
+	case 3:
+	{
+		g = new osg::PrimitiveSet(osg::PrimitiveSet::DrawElementsUBytePrimitiveType, (int)mode);
+		break;
+	}
+	case 4:
+	{
+		g = new osg::PrimitiveSet(osg::PrimitiveSet::DrawElementsUShortPrimitiveType, (int)mode);
+		break;
+	}
+	case 5:
+	{
+		g = new osg::PrimitiveSet(osg::PrimitiveSet::DrawElementsUIntPrimitiveType, (int)mode);
+		break;
+	}
+	default:
+	{
+		g = new osg::PrimitiveSet(osg::PrimitiveSet::PrimitiveType, (int)mode);
+		break;
+	}
+	}
+
+	if(g!=NULL)
+		g->ref();
+    return reinterpret_cast<jlong>(g);
+}
+*/
+
+JNIEXPORT jlong JNICALL Java_org_openscenegraph_osg_core_PrimitiveSet_nativeAsPrimitiveSet(JNIEnv *, jclass, jlong cptr)
+{
+	osg::PrimitiveSet *g = reinterpret_cast<osg::PrimitiveSet *>(cptr);
+	return reinterpret_cast<jlong>(g);
+}
```

Comment:
same as my other comment, asDrawable, asWhatever is a c++ smart way to avoid the dynamic_cast, in this case we only know about "jlong" pointers which are converted using reinterpret_cast, which forces the type, so.. better implement this at the Java level and remove it from here.